Audio device controls for mobile native audio device selection (#3270)
element-hq/element-call: docs/controls.md

# Global JS controls

A few aspects of Element Call's interface can be controlled through a global API on the `window`.

## Picture-in-picture

- `controls.canEnterPip(): boolean` Determines whether it's possible to enter picture-in-picture mode.
- `controls.enablePip(): void` Puts the call interface into picture-in-picture mode. Throws if not in a call.
- `controls.disablePip(): void` Takes the call interface out of picture-in-picture mode, restoring it to its natural display mode. Throws if not in a call.

## Audio output devices

These functions must be used in conjunction with the `controlledOutput` URL parameter in order to have any effect.

- `controls.setOutputDevices(devices: { id: string, name: string, forEarpiece?: boolean }[]): void` Sets the list of available audio outputs. `forEarpiece` is used on ios only.

>   It flags the device that should be used if the user selects earpice mode. This should be the main (stereo loudspeaker) of the device.

Reviewer: ```suggestion
  It flags the device that should be used if the user selects earpiece mode. This should be the main stereo loudspeaker of the device.
```

# Global JS controls

A few aspects of Element Call's interface can be controlled through a global API on the `window`.

## Picture-in-picture

- `controls.canEnterPip(): boolean` Determines whether it's possible to enter picture-in-picture mode.
- `controls.enablePip(): void` Puts the call interface into picture-in-picture mode. Throws if not in a call.
- `controls.disablePip(): void` Takes the call interface out of picture-in-picture mode, restoring it to its natural display mode. Throws if not in a call.

> ## Audio output devices

Reviewer: Same here

> These functions must be used in conjunction with the `controlledMediaDevices` URL parameter in order to have any effect.

Reviewer: Florian suggests this say:

> On mobile platforms (iOS, Android), web views do not reliably support selecting audio output devices such as the main speaker, earpiece, or headset. To work around this limitation, the following functions allow the hosting application (e.g., Element Web, Element X) to take control of audio device management via exposed JavaScript interfaces. These functions must be enabled by the URL parameter `controlledAudioDevices` in order to have any effect.

> - `controls.setAvailableOutputDevices(devices: { id: string, name: string, forEarpiece?: boolean, isEarpiece?: boolean isSpeaker?: boolean, isExternalHeadset?, boolean;}[]): void` Sets the list of available audio outputs. `forEarpiece` is used on ios only.

Reviewer: ```suggestion
- `controls.setAvailableOutputDevices(devices: { id: string, name: string, forEarpiece?: boolean, isEarpiece?: boolean isSpeaker?: boolean, isExternalHeadset?, boolean; }[]): void` Sets the list of available audio outputs. `forEarpiece` is used on iOS only.
```

  It flags the device that should be used if the user selects earpiece mode. This should be the main stereo loudspeaker of the device.
- `controls.onOutputDeviceSelect: ((id: string) => void) | undefined` Callback called whenever the user or application selects a new audio output.

> - `controls.setOutputDevice(id: string): void` Sets the selected audio device in EC menu. This should be used if the os decides to automatically switch to bluetooth.

Reviewer: ```suggestion
- `controls.setOutputDevice(id: string): void` Sets the selected audio device in Element Call's menu. This should be used if the OS decides to automatically switch to Bluetooth, for example.
```

Generally I'm trying to keep these docs in EC descriptive of how the APIs *could* be used rather than how downstream projects do happen to use them currently.

> - `controls.setOutputEnabled(enabled: boolean)` Enables/disables all audio output from the application. This can be useful for temporarily pausing audio while the controlling application is switching output devices. Output is enabled by default.

Reviewer: Just curious: do we still need this? I thought I remembered you saying that disabling audio output didn't actually help, in testing - not sure if I'm recalling that correctly

Author: Yes that is correct. We do have another possible usecase however.
Since we might want the audio mute button, there is a chance we want to enable the audio again if the user presses the hardware volume up button. Also we might want a "hardware down + lowest volume X times" to mute the audio.
This is a simple enough interface control this.
Once could still argue that this can be dropped from this pr.

Reviewer: Makes sense to me, I suggest we leave it in and just drop the sentence "This can be useful for temporarily pausing audio while the controlling application is switching output devices" then.

- `showNativeOutputDevicePicker: () => void`. This callback will be code by the webview if the user presses the output button in the settings menu.

>   This button is only shown on ios. (`userAgent.includes("IPhone")`)

Reviewer: ```suggestion
- `showNativeOutputDevicePicker: (() => void) | undefined`. Callback called whenever the user presses the output button in the settings menu.
  This button is only shown on iOS. (`userAgent.includes("iPhone")`)
```

To align with the other callback docs

> - `controls.setAudioEnabled(enabled: boolean)` Enables/disables all audio output from the application. This can be useful for temporarily pausing audio while the controlling application is switching output devices. Output is enabled by default.

Reviewer: ```suggestion
- `controls.setAudioEnabled(enabled: boolean)` Enables/disables all audio output from the application. Output is enabled by default.
```

I still suggest we drop this (since it won't be used that way after all)

Author: okay cool.
